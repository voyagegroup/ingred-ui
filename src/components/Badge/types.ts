import React from "react";
import { Theme } from "../../themes";

export type BadgeColor =
  | "primary"
  | "secondary"
  | "success"
  | "warning"
  | "danger"
  | "basic";

export type BadgeType = "normal" | "pill" | "signal";

export type BadgeSize = "medium" | "small";

// サイズの定数定義
export const BADGE_SIZE = {
  small: {
    iconSize: 14,
    height: "22px",
    pillHeight: "20px",
    gap: "2px",
    padding: (theme: Theme) => `${theme.spacing * 0.75}px`,
    signalPadding: (theme: Theme) => `${theme.spacing / 2}px`,
    signalGap: (theme: Theme) => `${theme.spacing / 2}px`,
    dotSize: "10px",
    fontSize: "12px",
  },
  medium: {
    iconSize: 16,
    height: "24px",
    pillHeight: "24px",
    gap: "4px",
    padding: (theme: Theme) => `${theme.spacing}px`,
    signalPadding: (theme: Theme) => `${theme.spacing * 0.75}px`,
    signalGap: (theme: Theme) => `${theme.spacing * 0.75}px`,
    dotSize: "12px",
    fontSize: "13px",
  },
};

// カラーマップの定義
export const BACKGROUND_COLOR_MAP = {
  normal: {
    primary: (theme: Theme) => theme.palette.primary.highlight,
    secondary: (theme: Theme) => theme.palette.basicDark.dark,
    success: (theme: Theme) =>
      theme.palette.success.softlight || theme.palette.success.light,
    warning: (theme: Theme) => theme.palette.warning.highlight,
    danger: (theme: Theme) => theme.palette.danger.highlight,
    basic: (theme: Theme) => theme.palette.black,
  },
  pill: {
    primary: (theme: Theme) => theme.palette.primary.main,
    secondary: (theme: Theme) => theme.palette.basicDark.light,
    success: (theme: Theme) => theme.palette.success.main,
    warning: (theme: Theme) => theme.palette.warning.main,
    danger: (theme: Theme) => theme.palette.danger.main,
    basic: (theme: Theme) => theme.palette.black,
  },
  signal: {
    primary: (theme: Theme) => theme.palette.primary.main,
    secondary: (theme: Theme) => theme.palette.gray.dark,
    success: (theme: Theme) => theme.palette.success.main,
    warning: (theme: Theme) => theme.palette.warning.main,
    danger: (theme: Theme) => theme.palette.danger.main,
    basic: (theme: Theme) => theme.palette.black,
  },
};

export const TEXT_COLOR_MAP = {
  normal: {
    primary: (theme: Theme) => theme.palette.text.primary,
    secondary: (theme: Theme) => theme.palette.text.secondary,
    success: (theme: Theme) => theme.palette.success.deepDark,
    warning: (theme: Theme) => theme.palette.warning.deepDark,
    danger: (theme: Theme) => theme.palette.danger.deepDark,
    basic: (theme: Theme) => theme.palette.text.white,
  },
  pill: {
    primary: (theme: Theme) => theme.palette.text.white,
    secondary: (theme: Theme) => theme.palette.text.secondary,
    success: (theme: Theme) => theme.palette.text.white,
    warning: (theme: Theme) => theme.palette.warning.deepDark,
    danger: (theme: Theme) => theme.palette.text.white,
    basic: (theme: Theme) => theme.palette.text.white,
  },
  signal: {
    primary: (theme: Theme) => theme.palette.text.primary,
    secondary: (theme: Theme) => theme.palette.text.secondary,
    success: (theme: Theme) => theme.palette.success.deepDark,
    warning: (theme: Theme) => theme.palette.warning.deepDark,
    danger: (theme: Theme) => theme.palette.danger.deepDark,
    basic: (theme: Theme) => theme.palette.text.primary,
  },
};

// 背景色の決定ロジック
export const getBackgroundColor = (
  key: BadgeColor,
  theme: Theme,
  type: BadgeType,
): string => {
  return BACKGROUND_COLOR_MAP[type][key](theme);
};

// テキスト色の決定ロジック
export const getTextColor = (
  key: BadgeColor,
  theme: Theme,
  type: BadgeType,
): string => {
  return TEXT_COLOR_MAP[type][key](theme);
};

export type BadgeProps = React.ComponentPropsWithoutRef<"a" | "span"> & {
  /**
   * バッジの色
   */
  color: BadgeColor;
  /**
   * バッジのタイプ
   * @default "normal"
   */
  type?: BadgeType;
  /**
   * フォントサイズ
   * @default "13px"
   */
  fontSize?: string;
  /**
   * フォントの太さ
   * @default "normal"
   */
  fontWeight?: string;
  /**
   * 使用するHTML要素
   * @default "span"
   */
  component?: "span" | "a";
  /**
   * バッジのサイズ
   * @default "medium"
   */
  size?: BadgeSize;
  /**
   * アイコン要素
   */
  icon?: React.ReactNode;
};
