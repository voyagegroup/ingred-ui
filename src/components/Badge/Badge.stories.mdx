import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import Badge from "./Badge";

<Meta title="Components/Data Display/Badge" component={Badge} />

# Badge

<ArgsTable of={Badge} />

## Samples

<Canvas>
  <Story
    name="primary"
    args={{
      color: "primary",
      children: "primary",
    }}
  >
    {(args) => <Badge {...args} />}
  </Story>
  <Story
    name="secondary"
    args={{
      color: "secondary",
      children: "secondary",
    }}
  >
    {(args) => <Badge {...args} />}
  </Story>
  <Story
    name="success"
    args={{
      color: "success",
      children: "success",
    }}
  >
    {(args) => <Badge {...args} />}
  </Story>
  <Story
    name="warning"
    args={{
      color: "warning",
      children: "warning",
    }}
  >
    {(args) => <Badge {...args} />}
  </Story>
  <Story
    name="danger"
    args={{
      color: "danger",
      children: "danger",
    }}
  >
    {(args) => <Badge {...args} />}
  </Story>
</Canvas>
