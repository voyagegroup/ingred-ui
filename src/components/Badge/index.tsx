export { default, BadgeProps } from "./Badge";
