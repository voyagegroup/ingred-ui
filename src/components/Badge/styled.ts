import styled from "styled-components";

export type BadgeType = "normal" | "pill";

type Props = {
  color: string;
  backgroundColor: string;
  type: BadgeType;
  fontSize: string;
  fontWeight: string;
};

export const Container = styled.span<Props>`
  display: inline-block;
  padding: ${({ theme }) => `${theme.spacing / 2}px ${theme.spacing}px`};
  border-radius: ${({ type, theme }) =>
    type === "pill" ? "10rem" : `${theme.radius}px`};
  background-color: ${({ backgroundColor }) => backgroundColor};
  text-align: center;
  font-size: ${({ fontSize }) => fontSize};
  font-weight: ${({ fontWeight }) => fontWeight};
  color: ${({ color }) => color};
  text-decoration: none;
`;
