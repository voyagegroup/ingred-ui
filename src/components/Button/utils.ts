import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
