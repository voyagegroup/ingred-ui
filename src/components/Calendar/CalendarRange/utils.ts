import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
