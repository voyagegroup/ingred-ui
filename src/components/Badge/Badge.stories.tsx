import React from "react";
import { StoryObj } from "@storybook/react";
import { Title, ArgsTable, Stories } from "@storybook/addon-docs";
import Badge from "./Badge";

export default {
  title: "Components/Data Display/Badge",
  components: Badge,
  parameters: {
    docs: {
      source: { language: "tsx" },
      page: () => (
        <>
          <Title />
          <ArgsTable of={Badge} />
          <Stories includePrimary title="Stories" />
        </>
      ),
    },
  },
};

const Template: StoryObj<typeof Badge> = {
  render: (args) => <Badge {...args} />,
};

export const Primary: StoryObj<typeof Badge> = {
  ...Template,
  args: {
    color: "primary",
    children: "primary",
  },
};

export const Secondary: StoryObj<typeof Badge> = {
  ...Template,
  args: {
    color: "secondary",
    children: "secondary",
  },
};
export const Success: StoryObj<typeof Badge> = {
  ...Template,
  args: {
    color: "success",
    children: "success",
  },
};

export const Warning: StoryObj<typeof Badge> = {
  ...Template,
  args: {
    color: "warning",
    children: "warning",
  },
};

export const Danger: StoryObj<typeof Badge> = {
  ...Template,
  args: {
    color: "danger",
    children: "danger",
  },
};

export const Basic: StoryObj<typeof Badge> = {
  ...Template,
  args: {
    color: "basic",
    children: "basic",
  },
};
