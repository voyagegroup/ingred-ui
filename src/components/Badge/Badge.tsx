import * as React from "react";
import * as Styled from "./styled";
import { Theme } from "../../themes";
import { useTheme } from "../../themes/useTheme";

export type BadgeColor =
  | "primary"
  | "secondary"
  | "success"
  | "warning"
  | "danger"
  | "basic";

const getColor = (key: BadgeColor, theme: Theme) => {
  switch (key) {
    case "primary":
      return theme.palette.primary.main;
    case "secondary":
      return theme.palette.gray.deepDark;
    case "success":
      return theme.palette.success.main;
    case "warning":
      return theme.palette.warning.main;
    case "danger":
      return theme.palette.danger.main;
    case "basic":
      return theme.palette.black;
  }
};

export type BadgeProps = React.ComponentPropsWithoutRef<"a" | "span"> & {
  color: BadgeColor;
  type?: Styled.BadgeType;
  fontSize?: string;
  fontWeight?: string;
  component?: "span" | "a";
};

const Badge = React.forwardRef<HTMLSpanElement | HTMLAnchorElement, BadgeProps>(
  function Badge(
    {
      color,
      type = "normal",
      component = "span",
      fontSize = "0.65em",
      fontWeight = "bold",
      children,
      ...rest
    },
    ref,
  ) {
    const theme = useTheme();
    return (
      <Styled.Container
        ref={ref}
        as={component}
        type={type}
        color={theme.palette.text.white}
        backgroundColor={getColor(color, theme)}
        fontSize={fontSize}
        fontWeight={fontWeight}
        {...rest}
      >
        {children}
      </Styled.Container>
    );
  },
);

export default Badge;
