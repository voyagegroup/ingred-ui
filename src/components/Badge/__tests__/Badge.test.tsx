import * as React from "react";
import "@testing-library/jest-dom";
import { cleanup } from "@testing-library/react";
import Badge from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Badge component testing", () => {
  afterEach(cleanup);

  test("Badge normal", () => {
    const { asFragment } = renderWithThemeProvider(
      <Badge color="primary">text</Badge>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("Badge pill", () => {
    const { asFragment } = renderWithThemeProvider(
      <Badge color="primary" type="pill">
        text
      </Badge>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
