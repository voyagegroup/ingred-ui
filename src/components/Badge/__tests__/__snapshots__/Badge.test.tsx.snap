// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge component testing Badge normal 1`] = `
<DocumentFragment>
  <span
    class="sc-bdvvtL jwswGa"
    color="#FFFFFF"
    font-size="0.65em"
    font-weight="bold"
    type="normal"
  >
    text
  </span>
</DocumentFragment>
`;

exports[`Badge component testing Badge pill 1`] = `
<DocumentFragment>
  <span
    class="sc-bdvvtL bVoBYZ"
    color="#FFFFFF"
    font-size="0.65em"
    font-weight="bold"
    type="pill"
  >
    text
  </span>
</DocumentFragment>
`;
